app/src/main/java/com/sschuraytz/charcoaldrawing/MainActivity.java

```diff
@@ -131,4 +158,121 @@ public void onUndoAvailable(boolean isAvailable) {
+
+            }
+
+            @Override
+            public void onEndOfSpeech() {
+                fab.setImageResource(R.drawable.ic_mic_foreground);
+            }
+
+            @Override
+            public void onError(int error) {
+                fab.setImageResource(R.drawable.ic_mic_foreground);
+            }
+
+            //Is there a way to make it more flexible so if user doesn't say exact word it still works?
+            @Override
+            public void onResults(Bundle results) {
+                ArrayList<String> data = results.getStringArrayList(SpeechRecognizer.RESULTS_RECOGNITION);
+                assert data != null;
+                String result = data.get(0);
+                if (result.equalsIgnoreCase("charcoal")) {
+                    Toast.makeText(getApplicationContext(), result, Toast.LENGTH_SHORT).show();
+                    drawingView.setDrawingMode();
+                }
+                else if (result.equalsIgnoreCase("eraser")) {
+                    Toast.makeText(getApplicationContext(), result, Toast.LENGTH_SHORT).show();
+                    drawingView.setEraseMode();
+                }
+                else if (result.equalsIgnoreCase("undo")) {
+                    Toast.makeText(getApplicationContext(), result, Toast.LENGTH_SHORT).show();
+                    drawingView.undo();
+                }
+                else if (result.equalsIgnoreCase("redo")) {
+                    Toast.makeText(getApplicationContext(), result, Toast.LENGTH_SHORT).show();
+                    drawingView.redo();
+                }
+                else if (result.equalsIgnoreCase("new")) {
+                    Toast.makeText(getApplicationContext(), result, Toast.LENGTH_SHORT).show();
+                    drawingView.createNewCanvas();
+                }
+                else {
+                    Toast.makeText(getApplicationContext(), "no such command", Toast.LENGTH_SHORT).show();
+                }
+                //smudge
+                //save
+                //help
+                //slider/radius - check if contains digit, if so, adjust slider
+            }
+
+            @Override
+            public void onPartialResults(Bundle partialResults) {
+
+            }
+
+            @Override
+            public void onEvent(int eventType, Bundle params) {
+
+            }
+        };
+    }
+    @RequiresApi(api = Build.VERSION_CODES.M)
+    public void listenToUserCommand()
+    {
+        //show user that mic is active
+        fab.setColorFilter(Color.RED);
+        //intent = simple message to transfer data btwn activities
+        Intent intent = new Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH);
+        intent.putExtra(RecognizerIntent.EXTRA_LANGUAGE_MODEL, "en-US");
+      //  intent.putExtra(RecognizerIntent.EXTRA_LANGUAGE_MODEL, RecognizerIntent.LANGUAGE_MODEL_FREE_FORM);
+        intent.putExtra(RecognizerIntent.EXTRA_MAX_RESULTS, 1);
+        speechRecognizer.startListening(intent);
```

Comment:
Do you ever call "stopListening()"?

Reply:
No. The recognizer stops automatically when it determines that the user has stopped speaking.
_"public void stopListening ()
Stops listening for speech. Speech captured so far will be recognized as if the user had stopped speaking at this point. Note that in the default case, this does not need to be called, as the speech endpointer will automatically stop the recognizer listening when it determines speech has completed."_ (https://developer.android.com/reference/android/speech/SpeechRecognizer)